### PR TITLE
[WIP] Introduce SQL sandbox

### PR DESCRIPTION
**Pull request by fishcakez**
@josevalim thoughts?

### CODE REVIEW

**Comment by josevalim** [comment]
This is beautiful. Can we merge this? :D

File: lib/ecto/adapters/sql/sandbox.ex [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+    true
+  end
+
+  defp demonitor(%{monitor: mon} = s) do
+    Process.demonitor(mon, [:flush])
+    %{s | fun: nil, ref: nil, monitor: nil}
+  end
+
+  defp dequeue(%{queries: queries} = s) do
+    case :queue.out(queries) do
```
Comment (josevalim) [inline lib/ecto/adapters/sql/sandbox.ex]: Question: why do we need the queue when we didn't need one with a pool of size 1?

Comment (fishcakez) [inline lib/ecto/adapters/sql/sandbox.ex]: We did have a queue: `poolboy`. If an application is using the single connection, the setup in an `ExUnit.Case` gets queued on `poolboy` waiting for the application to hand the single connection back in.

Should probably merge the queues. Having two different queues will reorder requests which might cause race conditions.

Comment (josevalim) [inline lib/ecto/adapters/sql/sandbox.ex]: Would it make more sense then to still have it behind poolboy? We need to ship with a default queue... so we may as well use it?

File: lib/ecto/adapters/sql/sandbox.ex [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+      {_, {:error, _} = error} ->
+        error
+    end
+  end
+
+  defp do_checkout(pool, ref, fun, timeout) do
+    try do
+      GenServer.call(pool, {:checkout, ref, fun}, timeout)
+    catch
+      :exit, {:timeout, _} = reason ->
```
Comment (fishcakez) [inline lib/ecto/adapters/sql/sandbox.ex]: Should catch `{:noproc, _}` too.

Comment (fishcakez) [inline lib/ecto/adapters/sql/sandbox.ex]: oh never mind we do :)

File: lib/ecto/adapters/sql/sandbox.ex (hunk 1 above)
Comment (fishcakez) [inline lib/ecto/adapters/sql/sandbox.ex]: Lets try behind poolboy - separate commit as discussed yesterday - before merging?

Comment (josevalim) [inline lib/ecto/adapters/sql/sandbox.ex]: Yes.

File: mix.lock [hunk 3]
```diff
@@ -1,4 +1,5 @@
-%{"decimal": {:hex, :decimal, "1.1.0"},
+%{"dbg": {:hex, :dbg, "1.0.1"},
```
Comment (fishcakez) [inline mix.lock]: oops

**Comment by josevalim** [comment]
@fishcakez I have merged only the first commit because the code we removed with queue handling from poolboy become pool management, so it is not really worth it.

If you prefer the poolboy version, let me know, and i will pick the other commit too. Thanks a lot!  :heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by fishcakez** [comment]
Theres bug in break when :raw

**Pull request closed without merging**